### ISSUE TITLE
[question] Using fst with giant Vector-like structures (integer keys & values)

### ISSUE THREAD

**Issue by nyurik** [BurntSushi/fst#125]
Hi, we are exploring ways for efficiently storing OpenStreetMap node index.  Each node has a sequentially growing u64 ID, and 2 i32 (latitude/longitude).  When processing OSM data, the most common operation is to look up latitude/longitude by their node ID. Currently there are [6,896,148,872 nodes](https://www.openstreetmap.org/stats/data_stats.html), and the last node just created has an ID 8,681,197,352. New nodes are constantly added, and some of the old ones are infrequently deleted.  Storing such data raw with the node ID would require extra 100% of space (64bit for lat+lng, plus 64bit for the ID), yet it doesn't handle deleted nodes, and as we can see from above numbers, that's wasted 20%.  The total "raw" memmapped index without node IDs takes about 65GB.

Given the above, do you think it would be possible to utilize fst for such storage?  Thanks!

**Comment by BurntSushi** [comment]
Hiya! I'd be happy to try and answer questions. But I think I have some of my own. I'm not quite sure I understand a few things.

> Storing such data raw with the node ID would require extra 100% of space (64bit for lat+lng, plus 64bit for the ID), yet it doesn't handle deleted nodes, and as we can see from above numbers, that's wasted 20%. The total "raw" memmapped index without node IDs takes about 65GB.

This is the main part that I don't understand. I think my confusion comes from not knowing to which thing you're referencing. For example, you say, "storing such data raw with the node ID would require extra 100% space." Do you mean in the FST? Do you mean some other representation? You're existing representation? I think it would help to be really specific here.

My guess is that your "raw" representation is just a sequence of endian-encoded 32-bit integers. Is that correct?

> New nodes are constantly added, and some of the old ones are infrequently deleted.

I can't quite tell how much you know about FSTs and this crate, but _this_ is where your problems will come from. FSTs are immutable. Once built, you cannot add or remove things from it. Therefore, you have to implement that part yourself. The "typical" way to do this is _conceptually_ simple:

* Deletions are implemented via tombstones. Lookups need only check if the node ID has been deleted by consulting the tombstone set before looking in the FST(s).
* Additions are handled by batching them, and then creating a new FST. All subsequent searches now require consulting the new FST along with any previously created FSTs.
* Since the addition of FSTs and tombstones will, over time, create unacceptable latency penalties in future lookups, you need to implement some form of periodic compaction process. This compaction process selects some FSTs to merge, and in the process of merging, removes entries from the tombstone map.

While conceptually simple, this is actually a fair bit of work to implement. And that amount of work depends directly on your access patterns and consistency guarantees. I could talk more about that if you have further questions.

**Comment by nyurik** [comment]
@BurntSushi thank you so much for an in-depth and quick reply!  Yes, i was referring to the raw endian encoded values for my current implementation -- essentially the whole thing is a giant mem-mapped array of u64 values, with the node ID being the index. It is possible to use the index in a read-only use case --  parse the whole planet data and use it, without subscribing to the real time updates. It would be interesting if fst could be used for the read-only case, but its usefulness would be severely diminished because  the common use case is to subscribe, i.e. to modify the data.

**Comment by BurntSushi** [comment]
Yes, I suspect `fst` would be excellent for the read-only use case, especially if you could put the file on a ramdisk or at least an SSD.

And yup, adding mutation to FSTs via the process I described above is a significant engineering effort. Mutating the FST itself isn't really an option unfortunately. It would require a totally different architecture with different trade offs. (And probably wouldn't be worth doing, I suspect.)

**Comment by nyurik** [comment]
Got it, thanks @BurntSushi . So for `u64 -> u64` map using fst, where would I get started?  All your examples use strings as keys, so I was a bit confused if this would be an efficient storage pattern.

**Comment by BurntSushi** [comment]
You would want to encode your `u64` keys as binary big-endian. That can be done easily with [`u64::to_be_bytes`](https://doc.rust-lang.org/std/primitive.u64.html#method.to_be_bytes). Note that big-endian here is important: it preserves the lexicographic ordering of the corresponding integer value. e.g., given integers `x` and `y`, and without loss of generality, you have that if `x < y` then `big-endian-bytes(x) < big-endian-bytes(y)`, where the former `<` is integer comparison and the latter is lexicographic comparison. The same property does not hold for little-endian.

And you'll need to insert your keys in lexicographic order as well.

**Comment by nyurik** [comment]
Once again, huge thanks for the quick and thorough reply!
